### ISSUE TITLE
Repeating event only showing up once (google .ics)

### ISSUE THREAD

**Issue by dwrolvink** [wernerjoss/grav-plugin-fullcalendar#8]
I made a new test calendar on calendar.google.com, and made one test event on date x, repeating weekly. In google, it shows up repeating indefinitely. Then I exported the .ics, and loaded it in using 

`[fullcalendar icsfile="http://mysite.com/user/data/calendars/test.ics"][/fullcalendar]`

This shows the event on date x, but not the weeks after that.

For reference, this is the ical file:

```
BEGIN:VCALENDAR
PRODID:-//Google Inc//Google Calendar 70.9054//EN
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
X-WR-CALNAME:test
X-WR-TIMEZONE:Europe/Amsterdam
BEGIN:VTIMEZONE
TZID:Europe/Amsterdam
X-LIC-LOCATION:Europe/Amsterdam
BEGIN:DAYLIGHT
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
DTSTART;TZID=Europe/Amsterdam:20190918T200000
DTEND;TZID=Europe/Amsterdam:20190918T220000
RRULE:FREQ=WEEKLY;BYDAY=WE
DTSTAMP:20190913T221236Z
UID:[redacted]rsolq@google.com
CREATED:20190913T221158Z
DESCRIPTION:
LAST-MODIFIED:20190913T221158Z
LOCATION:Repetitie
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:test
TRANSP:OPAQUE
END:VEVENT
END:VCALENDAR
```
Any idea what's going wrong here? Or is this not supported (yet)?

**Comment by dwrolvink** [comment]
OK, so I looked through the code and found the answer quicker than I thought.
It seems to not be supported yet. I started on a quick hack, but I lack some knowhow here.

Below I'll post the code that should be extended, and then below it how I have extended it.

**grav-plugin-fullcalendar/templates/partials/fullcalendar.html.twig**
```js
 //	map them to FullCalendar events
events = jQuery.map(eventComps, function(item) {
	if (item.getFirstPropertyValue("class") == "PRIVATE") {
		return null;
	} else {
		return {
			"title": item.getFirstPropertyValue("summary"), // + ";",
			"start": item.getFirstPropertyValue("dtstart").toJSDate(),
			"url": item.getFirstPropertyValue("url"),
			"end": item.getFirstPropertyValue("dtend").toJSDate(),
			"location": item.getFirstPropertyValue("location")
		};
	}
})
jQuery.merge(allevents, events);
```
---

**grav-plugin-fullcalendar/templates/partials/fullcalendar.html.twig**
```js
// save all events that repeat indefinitely
let repeats = [];

// add each item in the ical to the calendar
eventComps.forEach(function(item){
	if (item.getFirstPropertyValue("class") == "PRIVATE") {
		return null;
	}
	else {
		// add to calendar
		allevents[allevents.length] = {
			"title": item.getFirstPropertyValue("summary"), // + ";",
			"start": item.getFirstPropertyValue("dtstart").toJSDate(),
			"url": item.getFirstPropertyValue("url"),
			"end": item.getFirstPropertyValue("dtend").toJSDate(),
			"location": item.getFirstPropertyValue("location")
		};

		// save all events that repeat indefinitely
		rrule = item.getFirstPropertyValue("rrule");
		if (rrule && rrule.freq == "WEEKLY"){
			day = rrule.parts.BYDAY[0];
			repeats[repeats.length] = {
				"start": item.getFirstPropertyValue("dtstart").toJSDate(),
				"end": item.getFirstPropertyValue("dtend").toJSDate(),
				"freq": {"weekly": true, "day": rrule.parts.BYDAY[0]},
				"title": item.getFirstPropertyValue("summary"),
				"url": item.getFirstPropertyValue("url"),
				"location": item.getFirstPropertyValue("location")
			}
		}				
	}
})

// create new events for repeat events
let today = new Date();
for (i=0;i<120;i++){
	// loop through (today - 30) until (today + 30)
	// this is a hack, it would be better to request the current month being
	// shown from the calendar, and just make duplicate events for those dates
	// but alas, I don't know how...
	day = new Date(today.getTime());
	day.setDate(day.getDate() - 30 + i);
	
	// foreach day, check each event if it has a "duplicate event" on this day
	repeats.forEach(function(repeatEvent){
		if (repeatEvent.start.getTime() > day.getTime()){
			// first event from the series hasn't happened yet
			return;
		}
		
		// case: weekly repeat
		if (repeatEvent.freq.weekly){
			if (repeatEvent.start.getDay() == day.getDay() ){
				dif = repeatEvent.end.getDate() - repeatEvent.start.getDate(); 
				end = new Date(day.getTime()); end.setDate(end.getDate() + dif);
				
				// add repeat event
				allevents[allevents.length] = {
					"title": repeatEvent.title, // + ";",
					"start": day,
					"url": repeatEvent.url,
					"end": end,
					"location": repeatEvent.location
				};										
			} 
		}
	})

}														
```

So yeah, big mess. I wrote in the code comments: if someone knows how to request the current month that's being shown, then I can make it a lot less hacky. 

But even then, as it stands, I'm not good enough at JS to really clean this up nicely :)

**Comment by wernerjoss** [comment]
> OK, so I looked through the code and found the answer quicker than I thought.
> It seems to not be supported yet. 

that's right - repeated events are currently not supported.
I'll have a look at this later, when I have more time...

**Comment by wernerjoss** [comment]
I just searched the docs on fullcalendar.io for recurring events and found [this](https://fullcalendar.io/docs/recurring-events) - so, in short, rrules from ics calendars are not directly supported, but there are possibilities to implement that, simple recurrence or the rrule plugin.
ATM, I don't have the time to dig into this (ok, more important: i don't need it :D ) - maybe anyone else can do this and send a patch (or clone and create a pull request) ?

**Comment by wernerjoss** [comment]
there is a new Version (0.2.0) in trunk, which supports recurring events - see the docs (README.md).
But, as this is a major rewrite, also based on fullcalendar.io V 4.3.1 instead of 3.1, I have not yet created a release which will be distributed via the Grav Plugin updater, but rather would like to have this tested and confirmed working as expected by interested users.
Doing this is rather easy:

1. delete ALL files and directories in your user/plugins/fullcalendar folder (you can also uninstall the plugin, but that will also remove the fullcalendar folder, which will have to be recreated afterwards)
2. download the zip file for the current development version from here, unpack it somewhere on your computer
3. copy all files and folders from the unpacked archive to your user/plugins/fullcalendar folder
4. evtl. make Adaptions (locale: xx) to your user/config/plugins/fullcalendar.yaml file
5. clear cache

If anything goes wrong, you can always just delete your user/plugins/fullcalendar folder and then re-install to previous Version (0.1.5)
As always, comments, requests and patches are welcome :)

**Comment by wernerjoss** [comment]
closing as resolved, see issue #10 .
